modules/paths/src/main/java/coursier/paths/CachePath.java

```diff
@@ -93,7 +93,22 @@ public static File defaultCacheDirectory() throws IOException {
         return CoursierPaths.cacheDirectory();
     }
 
-    private static ConcurrentHashMap<Path, Object> processStructureLocks = new ConcurrentHashMap<Path, Object>();
+    // Trying to limit the calls to String.intern via this map (https://shipilev.net/jvm/anatomy-quarks/10-string-intern/)
+    private static ConcurrentHashMap<String, Object> internedStrings = new ConcurrentHashMap<>();
+
+    // Even if two versions of that code end up in the same JVM (say one via a shaded coursier, the other via a
+    // non-shaded coursier), they will rely on the exact same object for locking here (via String.intern), so that the
+    // locks are actually JVM-wide.
+    private static Object lockFor(Path cachePath) {
```

Comment:
It might make sense to synchronize this method to be extra safe.

Reply:
We're only relying on thread-safe methods in it ([`String.intern`](https://stackoverflow.com/questions/14798003/is-string-intern-thread-safe), `ConcurrentHashMap.putIfAbsent`, …), it should be ok. Plus locks aren't free.

Reply:
I was thinking about what the jdk does for the classloading lock: https://github.com/openjdk/jdk/blob/439a48bd1cdaec0337f296102752c4253e1d944c/src/java.base/share/classes/java/lang/ClassLoader.java#L663
I guess the use of interned strings is more or less equivalent since you're guaranteed to get the same object out of key.intern()? Otherwise I could see a race condition where two threads concurrently intern the string and synchronize on a different instance. I'm not saying this is wrong since I don't fully understand it, but it would be more obviously correct if it was:
```
lock0 = internedStrings.putIfAbsent(internedKey, internedKey);
lock0 = lock0 == null ? internedKey : lock0;
```